### PR TITLE
[OPIK-275] fix bubbling 500 error from EM

### PR DESCRIPTION
**Pull request by idoberko2**
## Details
Fix bubbling 500 error from EM backend and added some coverage.

## Issues
OPIK-275

## Testing
Added some test

### CODE REVIEW

File: apps/opik-backend/src/main/java/com/comet/opik/infrastructure/auth/RemoteAuthService.java [hunk 1]
```diff
@@ -145,13 +145,12 @@ private AuthResponse verifyResponse(Response response) {
...
         } else if (response.getStatus() == Response.Status.UNAUTHORIZED.getStatusCode()) {
-            throw new ClientErrorException(NOT_ALLOWED_TO_ACCESS_WORKSPACE,
-                    Response.Status.UNAUTHORIZED);
+            throw new ClientErrorException(NOT_ALLOWED_TO_ACCESS_WORKSPACE, Response.Status.UNAUTHORIZED);
         } else if (response.getStatus() == Response.Status.FORBIDDEN.getStatusCode()) {
             throw new ClientErrorException("User has bot permission to the workspace", Response.Status.FORBIDDEN);
         } else if (response.getStatusInfo().getFamily() == Response.Status.Family.SERVER_ERROR) {
             log.error("Error while authenticating user");
-            throw new ClientErrorException(Response.Status.INTERNAL_SERVER_ERROR);
+            throw new ClientErrorException(Response.Status.BAD_REQUEST);
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/infrastructure/auth/RemoteAuthService.java]: There's another potential occurrence of this below. Please fix it as well.

File: apps/opik-backend/src/test/java/com/comet/opik/infrastructure/auth/RemoteAuthTestServer.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import lombok.Setter;
+import org.apache.hc.core5.http.HttpStatus;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+
+@NoArgsConstructor
+class RemoteAuthTestServer {
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/infrastructure/auth/RemoteAuthTestServer.java]: I'm not against this approach, but probably it's more lightweight to create a mock resource, that can take Mockito mocks and use the dropwizard extension.

**Review by andrescrz — COMMENTED** [review]
Let's fix the other occurence.

File: apps/opik-backend/src/test/java/com/comet/opik/infrastructure/auth/RemoteAuthTestServer.java (hunk 2 above)
Comment (thiagohora) [reply apps/opik-backend/src/test/java/com/comet/opik/infrastructure/auth/RemoteAuthTestServer.java]: We can use WireMock for it

Comment (amirzatcomet) [inline apps/opik-backend/src/test/java/com/comet/opik/infrastructure/auth/RemoteAuthTestServer.java]: Technically I'm not against writing tests, but why was it important to add these tests right now?  what are we protecting here?

File: apps/opik-backend/src/main/java/com/comet/opik/infrastructure/auth/RemoteAuthService.java (hunk 1 above)
Comment (andrescrz) [reply apps/opik-backend/src/main/java/com/comet/opik/infrastructure/auth/RemoteAuthService.java]: You should do the reverse fix. 

From the external user perspective, if this internal call towards EM failed with a `SERVER_ERROR` we should still return an `INTERNAL_SERVER_ERROR` to the user, not a `BAD_REQUEST`. 

But that should be the legitimate 500 code, instead of a 500 coming because instantiating the exception object here throws a `IllegalArgumentException` due to the miss-match  between the `ClientErrorException` and the former `INTERNAL_SERVER_ERROR`:
- https://github.com/jakartaee/rest/blob/49b07656ebdb60b3386d0b542f1c2151961ec8a0/jaxrs-api/src/main/java/jakarta/ws/rs/WebApplicationException.java#L270-L272

In summary, the reverse fix should be changing this to an `InternalServerErrorException`.

Finally, the fact that there are two occurrences of this problem is caused by some previous spaghetti code. Basically this last `else if` statement is the same as the code below outside the `if-else` chain

You should remove that duplication, either leave one last `else` statement or leave the code outside the `if-else` chain, but not both. Basically, we just need one piece of code as "catch-all" internal server exception, not two pieces doing the same.

**Review by andrescrz — CHANGES_REQUESTED** [review]
Noticed that this is the wrong fix.

File: apps/opik-backend/src/main/java/com/comet/opik/infrastructure/auth/RemoteAuthService.java (hunk 1 above)
Comment (thiagohora) [reply apps/opik-backend/src/main/java/com/comet/opik/infrastructure/auth/RemoteAuthService.java]: I was going to ask this why 400 instead of 500

Comment (andrescrz) [reply apps/opik-backend/src/main/java/com/comet/opik/infrastructure/auth/RemoteAuthService.java]: The JIRA ticket title is misleading as it says:

> When workspace is absent authentication is returning 500

That's correct in my view, but for that first we need to improve the real Authentication endpoint to return a 4XX in those cases, so this endpoint also propagates a 4XX.

Then, there's the underlaying issue on this PR to fix the bubbling up IllegalArgumentException that ends up in a 5XX, by still returning a 5XX, but coming from a InternalServerErrorException.

File: apps/opik-backend/src/test/java/com/comet/opik/infrastructure/auth/RemoteAuthTestServer.java (hunk 2 above)
Comment (idoberko2) [reply apps/opik-backend/src/test/java/com/comet/opik/infrastructure/auth/RemoteAuthTestServer.java]: changed to wiremock in https://github.com/comet-ml/opik/pull/754/commits/57dea2355bdcb6cf1187fb28759d8b37d2c70fae
I wasn't aware it has this feature set

Comment (idoberko2) [reply apps/opik-backend/src/test/java/com/comet/opik/infrastructure/auth/RemoteAuthTestServer.java]: This seems like a critical path to me, which has low probability to change in the near future and worth covering with tests when I'm inside the context anyway

File: apps/opik-backend/src/main/java/com/comet/opik/infrastructure/auth/RemoteAuthService.java (hunk 1 above)
Comment (idoberko2) [reply apps/opik-backend/src/main/java/com/comet/opik/infrastructure/auth/RemoteAuthService.java]: Agree with all of the above. Fixed.

**Comment by amirzatcomet** [comment]
This is a minor fix, I think we've spent enough time on it, let's finish up and move on.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/WireMockUtils.java [hunk 3]
```diff
@@ -14,9 +14,18 @@ public record WireMockRuntime(WireMockRuntimeInfo runtimeInfo, WireMockServer se
...
-
         final WireMockServer wireMockServer = new WireMockServer(wireMockConfig().dynamicPort().dynamicHttpsPort());
 
+        return startWireMock(wireMockServer);
+    }
+
+    public static WireMockRuntime startWireMockHttpOnly() {
+        final WireMockServer wireMockServer = new WireMockServer(wireMockConfig().dynamicPort());
+
+        return startWireMock(wireMockServer);
```
Comment (thiagohora) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/WireMockUtils.java]: The previous server should give you both HTTP and https

**Review by thiagohora — DISMISSED** [review]

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/WireMockUtils.java (hunk 3 above)
Comment (idoberko2) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/WireMockUtils.java]: I'm receiving a certificate error for it. It works using http. If you have any idea how to make it work, I'm open for suggestions...

Comment (thiagohora) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/WireMockUtils.java]: Got it. You can use TestHttpClientUtils.client() The should solve your issue

Comment (idoberko2) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/WireMockUtils.java]: Done. Thanks!

**Review by thiagohora — APPROVED** [review]

**Review by andrescrz — APPROVED** [review]
LGTM.
